### PR TITLE
Reworked KVStore entry handling

### PR DESCRIPTION
**Pull request by krzlabrdx**

### CODE REVIEW

File: src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/KeyValueStoreProcessor.cs [hunk 1]
```diff
@@ -83,11 +83,12 @@ internal class KeyValueStoreProcessor
 {
     private readonly ProcessorContext _context;
     private ChangeTracker<KeyValueStoreChangePointerLookup, KeyValueStoreChangePointer> _changes = new();
-    private Dictionary<long, KeyValueStoreAggregateHistory> _mostRecentAggregates = new();
-    private Dictionary<KeyValueStoreEntryDbLookup, KeyValueStoreEntryHistory> _mostRecentEntries = new();
 
-    private List<KeyValueStoreAggregateHistory> _aggregatesToAdd = new();
-    private List<KeyValueStoreEntryHistory> _entriesToAdd = new();
+    private Dictionary<KeyValueStoreEntryDbLookup, long> _observedEntryDefinitions = new();
```
Comment (PawelPawelec-RDX) [inline src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/KeyValueStoreProcessor.cs:87]: all these propos can be readonly.

File: src/RadixDlt.NetworkGateway.PostgresIntegration/CommonDbContext.cs [hunk 2]
```diff
@@ -530,12 +530,16 @@ private static void HookupHistory(ModelBuilder modelBuilder)
             .HasIndex(e => e.Epoch);
 
         modelBuilder
-            .Entity<KeyValueStoreEntryHistory>()
-            .HasIndex(e => new { e.KeyValueStoreEntityId, e.Key, e.FromStateVersion });
+            .Entity<KeyValueStoreEntryDefinition>()
+            .HasIndex(e => new { e.KeyValueStoreEntityId, e.FromStateVersion });
```
Comment (PawelPawelec-RDX) [inline src/RadixDlt.NetworkGateway.PostgresIntegration/CommonDbContext.cs:534]: I think we're missing one index for that query/filter that would include kv entity id, key, from_state_version

`WHERE d.key_value_store_entity_id = @keyValueStoreEntityId AND d.key = ANY(@keys) AND d.from_state_version <= @stateVersion"`

File: src/RadixDlt.NetworkGateway.PostgresIntegration/Services/KeyValueStoreQuerier.cs [hunk 3]
```diff
@@ -98,68 +97,70 @@ public KeyValueStoreQuerier(IDapperWrapper dapperWrapper, ReadOnlyDbContext dbCo
...
-    kvseh.key AS Key,
-    kvsits.key_value_store_items_total_count AS TotalCount
-FROM key_value_store_items_slices AS kvsits
-INNER JOIN LATERAL UNNEST(key_value_store_items_slice) WITH ORDINALITY AS key_value_store_join(id, ordinality) ON TRUE
-INNER JOIN key_value_store_entry_history kvseh ON kvseh.id = key_value_store_join.id AND kvseh.is_deleted = FALSE
-ORDER BY key_value_store_join.ordinality ASC
+) h ON TRUE
+WHERE
+    d.key_value_store_entity_id = @keyValueStoreEntityId
+  AND (d.from_state_version, d.id) <= (@cursorStateVersion, @cursorId)
```
Comment (PawelPawelec-RDX) [inline src/RadixDlt.NetworkGateway.PostgresIntegration/Services/KeyValueStoreQuerier.cs:123]: I'm not sure but maybe providing cursorStateVersion and cursorId as nullable to sql and rework to not apply filters if they are null will be more performant?

File: src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/PostgresLedgerExtenderService.cs [hunk 4]
```diff
@@ -224,6 +224,7 @@ private async Task<ExtendLedgerReport> ProcessTransactions(ReadWriteDbContext db
         var outerStopwatch = Stopwatch.StartNew();
         var referencedEntities = new ReferencedEntityDictionary();
         var referencedNonFungibleIdDictionary = new ReferencedNonFungibleIdDictionary();
+        var referencedKeyValueStoreEntries = new ReferencedKeyValueStoreEntryDictionary();
```
[the remark this answers is not in the document]
Comment (krzlabrdx) [reply src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/PostgresLedgerExtenderService.cs]: Dropped.

File: src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/ReferencedKeyValueStoreEntryDictionary.cs [hunk 5]
```diff
@@ -62,91 +62,40 @@
...
+namespace RadixDlt.NetworkGateway.PostgresIntegration.LedgerExtension;
 
-internal record KeyValueStoreChange(long KeyValueStoreEntityId, long StateVersion, KeyValueStoreEntry Entry);
+internal record KeyValueStoreEntryLookup(EntityAddress KeyValueStoreAddress, ValueBytes Key);
 
-internal record KeyValueStoreEntry(ValueBytes Key, ValueBytes? Value);
+internal record KeyValueStoreEntryDefinitionDatabaseId(long KeyValueStoreEntityId, long KeyValueStoreEntryDefinitionId);
 
-internal static class KeyValueStoreExtensions
+internal class ReferencedKeyValueStoreEntryDictionary
```
[the remark this answers is not in the document]
Comment (krzlabrdx) [reply src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/ReferencedKeyValueStoreEntryDictionary.cs]: Dropped.

File: src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/KeyValueStoreProcessor.cs (hunk 1 above)
Comment (krzlabrdx) [reply src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/KeyValueStoreProcessor.cs]: Changed.

File: src/RadixDlt.NetworkGateway.PostgresIntegration/CommonDbContext.cs (hunk 2 above)
Comment (krzlabrdx) [reply src/RadixDlt.NetworkGateway.PostgresIntegration/CommonDbContext.cs:534]: I think we have all the necessary indices:
```
// used while iterating over KVStore entries
        modelBuilder
            .Entity<KeyValueStoreEntryDefinition>()
            .HasIndex(e => new { e.KeyValueStoreEntityId, e.FromStateVersion });

// used for key lookup
        modelBuilder
            .Entity<KeyValueStoreEntryDefinition>()
            .HasIndex(e => new { e.KeyValueStoreEntityId, e.Key });
```

File: src/RadixDlt.NetworkGateway.PostgresIntegration/Services/KeyValueStoreQuerier.cs (hunk 3 above)
Comment (krzlabrdx) [reply src/RadixDlt.NetworkGateway.PostgresIntegration/Services/KeyValueStoreQuerier.cs:123]: It results in extra `index cond` (`Filter: (ROW(from_state_version, id) <= ROW('11111111111'::bigint, '2222222222222'::bigint))"`) that should have negligible impact on the query.

**Review by PawelPawelec-RDX — APPROVED** [review]

**Pull request merged**